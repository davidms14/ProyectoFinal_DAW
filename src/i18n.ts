import i18n from 'i18next';
import { initReactI18next } from 'react-i18next';

/**
 * Este documento contiene las funciones para poder implementar la internalización en el proyecto 
 */

    //Idiomas a implementar
// English
import en from './assets/locales/en.json';

// Spanish
import es from './assets/locales/es.json';

export enum Language {
    English = 'en',
    Spanish = 'es'
}

//Array de palabras traducidas en el documento
const cldr = {
    [Language.English]: new Array(),
    [Language.Spanish]: new Array(),
};

const resources = {
    [Language.English]: {
        translation: en
    },
    [Language.Spanish]: {
        translation: es
    }
};

i18n
    .use(initReactI18next) // pasa i18n abajo a react-i18next
    .init({
        resources,
        lng: Language.English, // Cambia el lenguaje actual al final del documento
        interpolation: {
            escapeValue: false 
        },
    });


export default i18n;

export function getCurrentLanguage(): string {
    return i18n.language;
}

export type LanguageReturnType = string[];

export function getLanguagesList(): LanguageReturnType[] {
    return [
        ['English', Language.English],
        ['Español', Language.Spanish]
    ];
}

export function changeCurrentLanguage(language: Language): void {
    i18n.changeLanguage(language);

}

// Si no esta implementado el inglés por defecto, usamos el español
changeCurrentLanguage(Language.Spanish);