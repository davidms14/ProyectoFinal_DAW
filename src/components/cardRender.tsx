import '../index.css';
import '../style.css';
import * as React from 'react';
import { withTranslation, WithTranslation } from 'react-i18next';
import { getCurrentLanguage } from '../i18n';
import moment from 'moment';

/**
 * @param cardRender - Esta clase funciona como un componente para Tile ya que es la encargada 
 * de contener la infomación que se mostrará en las tarjetas, los datos de los clientes, fechas, muestras
 * de estas... Además tendrá unos iconos que nos permitirá conocer en todo momento el estado de las tarjetas
 * que estamos mostrando. Con estos iconos que ejercen las funciones de botón podremos visualizar un reporte
 * de manera individual, observar las muestras enlazadas a dicho reporte y descargar el reporte seleccionado.
 */

export interface Data {
    id: string;
    reportNumber: string;
    client: string;
    entryDate: Date;
    expectedDate: Date;
    status: string;
    reference: string;
    certificate: string;
}

interface Props {
    data: Data;
}

interface CardRenderState {
    data: Data;
    isLoadSamples: boolean;
    hasMore: boolean;
}

export type CardRenderPropsWithTranslation = Props & WithTranslation;

export class CardRender extends React.Component<CardRenderPropsWithTranslation, CardRenderState> {
    public constructor(props: CardRenderPropsWithTranslation) {
        super(props);

        this.state = {
            data: this.props.data,
            isLoadSamples: false,
            hasMore: true,
        };
    }

    /**
     * 
     * @returns {string} Devuelve string
     * @example
     * //Example funcion
     * @description
     * //Con esta función podemos controlar el estado y saber si
     * se ha acabado de trabajar en un reporte o sigue en proceso. Esto puede saberse grácias a los iconos
     * que nos ayudan a identificarlo.
     * @param {string} status - Hace refencia al estado del reporte.
     * 
     */
    private getStatus(status: string): string {
        if (status === "finalized") {
            /** Tarea finalizada */
            return status = "icon-checkbox-checked iconosTarjetas";
        }
        else {
            /** Tarea por finalizar */
            return status = "icon-hour-glass iconosTarjetas"
        }
    }

    /**
     * @returns {string} Devuelve string
     * @example
     * //Ejemplo funcion
     * @description
     * //Esta función simplemente permite saber el significado del icono que
     * nos indica el estado del reporte. Así al pasar el ratón por encima de este podemos saber 
     * que significa.
     * @param {string} title - Hace refencia al estado del reporte.
     */

    private getTitle(title: string): string {
        if (title === "finalized") {
            return title = this.props.i18n.t('status.finalized');
        }
        else {
            return title = this.props.i18n.t('status.inprogress');
        }
    }

    /**
     * @example
     * //Example funcion
     * @description
     * //La función se encarga de cargar las muestras, también modifica el estado de estas
     * ya que no siempre sacará las mismas, sino que deberá sacar la muestras enlazadas al reporte.
     * @param {boolean} isLoadSample - Controla si las muestras se han cargado
     */

    private getAlert(isLoadSample: boolean): void {
        /*var newState = this.state.data;*/
        this.setState({ /*data: newState,*/ isLoadSamples: isLoadSample });
    }


    public render(): JSX.Element {
        /** Almacena el contenido de la tarjeta, se mostrará información de la tarjeta o las muestras */
        let content = [];


        /** Antes se tomaban las muestras desde otro documento, ahora al ser un campo dentro del actual JSON accedemos a dicho campo
         * y lo metemos en un array. En cada posición se inserta una muestra, estas se separan por comas.
         */
        let references: string[] = this.state.data.reference.split(', ');
        let notEmptyReferences: string[] = references.filter((element: string): boolean => {
            return element !== undefined && element != null && element.trim() !== '';
        });

        /** Este icono nos mostrará las muestras, si estamos en ellas desaparece. */
        const buttonToShowSamples =
            (notEmptyReferences.length > 0 && !this.state.isLoadSamples) ?
                <i title={this.props.i18n.t('icons.samples')} className="icon-file-empty iconosTarjetas"
                    key={'buttonToShowSamples_'.concat(this.state.data.id)}
                    onClick={this.getAlert.bind(this, true)}
                >
                </i>
                :
                null;

        /** Con ese icono podremos visualizar el reporte */
        const buttonToShow =
            this.state.data.certificate === "true" ?
                <i title={this.props.i18n.t('icons.detail')} className="icon-link iconosTarjetas" ></i>
                :
                null;

        /** Con ese icono podremos descargar el reporte */
        const buttonToDownload =
            this.state.data.status === "finalized" ?
                <i title={this.props.i18n.t('icons.download')} className="icon-download3 iconosTarjetas" ></i>
                :
                null;

        /** Si estamos viendo las muestras, para volver a los datos del reporte pulsaremos aqui. */
        const buttonToClose =
            <i title={this.props.i18n.t('icons.cerrar')} className="icon-cross iconoCerrar"
                key={'buttonToCloseSamples_'.concat(this.state.data.id)}
                onClick={this.getAlert.bind(this, false)}
            ></i>;

        /** Contiene los datos a mostrar en la tarjeta */
        if (!this.state.isLoadSamples) {
            if (this.state.data.status === "finalized") {
                moment.locale(getCurrentLanguage());

                content.push(

                    <div key={'div_report_'.concat(this.state.data.id)}>
                        <b title={this.props.i18n.t('home.filters.client')}>{this.props.i18n.t('home.filters.client')}:</b> {this.state.data.client}
                        <br />
                        <b title={this.props.i18n.t('home.filters.entry')}>{this.props.i18n.t('home.filters.finish')}:</b> {moment(this.state.data.entryDate).format('L')}
                        <br />
                        <b title={this.props.i18n.t('home.filters.expected')}>{this.props.i18n.t('home.filters.timed')}:</b> {moment(this.state.data.expectedDate).add(5, 'years').format('L')}
                    </div>
                );
            } else {
                moment.locale(getCurrentLanguage());

                content.push(

                    <div key={'div_report_'.concat(this.state.data.id)}>
                        <b title={this.props.i18n.t('home.filters.client')}>{this.props.i18n.t('home.filters.client')}:</b> {this.state.data.client}
                        <br />
                        <b title={this.props.i18n.t('home.filters.entry')}>{this.props.i18n.t('home.filters.entry')}:</b> {moment(this.state.data.entryDate).format('L')}
                        <br />
                        <b title={this.props.i18n.t('home.filters.expected')}>{this.props.i18n.t('home.filters.expected')}:</b> {moment(this.state.data.expectedDate).format('L')}
                    </div>
                );
            }
        }

        /** Muestras los reportes asociados a la tarjeta*/
        else if (this.state.data.reference.trim() !== '') {
            content.push(buttonToClose);
            references.forEach((element: string, index: number): void => {
                content.push(
                    <div key={'div_reference_'.concat(index.toString())}>
                        <div><b>-</b> {element}</div>
                    </div>
                );
            });
        }

        /** Devuelve la esctrutura de la tarjeta, esta consta de un HEADER donde encontramos el estado y el número de reporte. 
         * Ambas cosas siempre estarán en ese sitio.
         * 
         * Seguidamente, tenemos el apartado de CONTENT, el cual si que varia dependiendo de si estamos mostrando la información
         * de la tarjeta o las muestras.
         * 
         * Para acabar tendriamos un apartado de ACTIONS, este contiene los tres botones de detalles, descarga y muestras y al igual que la cabecera
         * estos siempre ocuparán ese lugar.
          */
        return (
            <div className='e-card'>

                <div className='e-card-header'>
                    <div className='e-card-header-title' title={this.props.i18n.t('home.table.reportNumber')}>
                        <i title={this.getTitle(this.state.data.status)}
                            className={this.getStatus(this.state.data.status)}>
                        </i>{this.state.data.reportNumber}
                    </div>
                </div>

                <div className='e-card-content' style={{ overflowY: 'auto' }}>
                    {content}
                </div>

                <div className='e-card-actions iconos'>
                    {buttonToShow}
                    {buttonToShowSamples}
                    {buttonToDownload}
                </div>
            </div>
        );
    }
}


export default withTranslation()(CardRender);


