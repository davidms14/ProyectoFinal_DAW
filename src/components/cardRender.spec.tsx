import * as React from 'react';
import { shallow } from 'enzyme';
import { CardRender, Data, CardRenderPropsWithTranslation } from './cardRender';
import i18n from '../i18n';

/**
 * Documento para almacenar los tests del cardRender 
 */

 /**
  * Primer test: test para comprobar si se realiza la vista correcta de las tarjetas en forma de snapshot
  */

describe('components/cardRender tests', () => {
    it('should render as expected when passing required properties', () => {
        // Arrange
        const myData: Data = {
            id: '6fdd1b8f-b0b4-4b40-931f-fb95bb7e9b9a',
            reportNumber: '2019USAM17',
            client: 'testcreate',
            entryDate: new Date,
            expectedDate: new Date,
            status: 'inprogress',
            reference: 'TestAdd, TestAdd2, TestAdd3',
            certificate: 'true',
        };

        const props: CardRenderPropsWithTranslation = {
            data: myData,
            i18n: i18n
        } as CardRenderPropsWithTranslation;

        // Act
        const component = shallow(
            <CardRender {...props} />
        );

        // Assert
        expect(component.debug()).toMatchSnapshot();
    });

    /**
     * Segundo test: test para comprobar el funcionamiento de los botones de las tarjetas.
     */
    it('should try buttons in cards', () => {
        // Arrange
        const myData: Data = {
            id: '6fdd1b8f-b0b4-4b40-931f-fb95bb7e9b9a',
            reportNumber: '2019USAM17',
            client: 'testcreate',
            entryDate: new Date,
            expectedDate: new Date,
            status: 'inprogress',
            reference: 'TestAdd, TestAdd2, TestAdd3',
            certificate: '',
        };

        const props: CardRenderPropsWithTranslation = {
            data: myData,
            i18n: i18n
        } as CardRenderPropsWithTranslation;

        // Act
        const component = shallow(
            <CardRender {...props} />
        );

        // Assert
        component.find('.icon-file-empty').simulate('click');
        component.find('.icon-cross').simulate('click');
        expect(component.find('.icon-file-empty').length).toBe(1);
    });

    /**
     * Tercer test: Test para comprobar si estÃ¡ cargando la vista del reporte o la vista de las muestras
     */
    it('should render cards or samples', () => {
        // Arrange
        const myData: Data = {
            id: '6fdd1b8f-b0b4-4b40-931f-fb95bb7e9b9a',
            reportNumber: '2019USAM17',
            client: 'testcreate',
            entryDate: new Date,
            expectedDate: new Date,
            status: 'inprogress',
            reference: '',
            certificate: 'false'
        };

        const props: CardRenderPropsWithTranslation = {
            data: myData,
            i18n: i18n
        } as CardRenderPropsWithTranslation;

        // Act
        const component = shallow(
            <CardRender {...props} />
        );

        // Assert
        component.setState({ isLoadSamples: true });
        expect(1).toBe(1);
    });
});
