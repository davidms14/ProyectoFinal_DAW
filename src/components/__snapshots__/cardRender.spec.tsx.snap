// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/cardRender tests should render as expected when passing required properties 1`] = `
"<div className=\\"e-card\\">
  <div className=\\"e-card-header\\">
    <div className=\\"e-card-header-title\\" title=\\"NÂº Informe\\">
      <i title=\\"En Proceso\\" className=\\"icon-hour-glass iconosTarjetas\\" />
      2019USAM17
    </div>
  </div>
  <div className=\\"e-card-content\\" style={{...}}>
    <div>
      <b title=\\"Cliente\\">
        Cliente
        :
      </b>
       
      testcreate
      <br />
      <b title=\\"Fecha Inicio\\">
        Fecha Inicio
        :
      </b>
       
      05/06/2019
      <br />
      <b title=\\"Fecha estimada\\">
        Fecha estimada
        :
      </b>
       
      05/06/2019
    </div>
  </div>
  <div className=\\"e-card-actions iconos\\">
    <i title=\\"Ir a detalle\\" className=\\"icon-link iconosTarjetas\\" />
    <i title=\\"Mostrar muestras\\" className=\\"icon-file-empty iconosTarjetas\\" onClick={[Function: bound getAlert]} />
  </div>
</div>"
`;
